src/CloudFoundry.WinDEA/StagingInstance.cs

```diff
@@ -265,34 +267,226 @@ public void GetBuildpack(StagingStartMessageRequest message, string gitPath, str
+                        {
+                            Directory.Delete(Path.Combine(this.Workspace.TempDir, "buildpack"), true);
+                        }
                     }
                 }
 
+
                 if (this.Buildpack == null)
                 {
                     throw new Exception("Unable to detect a supported application type");
                 }
                 Logger.Info("Staging task {0}: Detected buildpack {1}", this.Properties.TaskId, this.Buildpack.Name);
             }
+
             this.Properties.DetectedBuildpack = this.Buildpack.Name;
         }
 
+        public static bool IsWindowsCompatibleBuildpack(string buildpackDir)
+        {
+            var bins = new string[] { "detect", "compile", "release" };
+            string[] pathExt = Environment.GetEnvironmentVariable("PATHEXT").Split(';');
+
+            foreach (var app in bins)
+            {
+                bool found = false;
+                foreach (string ext in pathExt)
+                {
+                    if (File.Exists(Path.Combine(buildpackDir, "bin", app + ext)))
+                    {
+                        found = true;
+                    }
+                }
+                if (!found)
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        public void DownloadAdminBuildpack(StagingStartRequestAdminBuildpack adminBuildpack, string adminBuildpackDir)
+        {
+            string archiveFile = Path.Combine(adminBuildpackDir, adminBuildpack.Key + ".zip");
+            string destDir = Path.Combine(adminBuildpackDir, adminBuildpack.Key);
+
+            if (Directory.Exists(destDir))
+            {
+                Logger.Info("Skipping download for admin buildpack {0} because it was previously downloaded in {1} ", adminBuildpack.Key, destDir);
+                return;
+            }
+
+            Directory.CreateDirectory(destDir);
+
+            WebClient client = new WebClient();
+
+            Logger.Info("Downloading admin buildpack {0} in {1} from {2}", adminBuildpack.Key, destDir, adminBuildpack.Url);
+            try
+            {
+                var downloadUri = new Uri(adminBuildpack.Url);
+                client.Headers[HttpRequestHeader.Authorization] = "Basic " + System.Convert.ToBase64String(System.Text.Encoding.UTF8.GetBytes(downloadUri.UserInfo));
+                client.DownloadFile(downloadUri, archiveFile);
+
+                ZipFile.ExtractToDirectory(archiveFile, destDir);
+            }
+            catch
+            {
+                Directory.Delete(destDir, true);
+                File.Delete(archiveFile);
+
+                throw;
+            }
+            finally
+            {
+                client.Dispose();
+            }
+
+            File.Delete(archiveFile);
+
+        }
+
+        public void TryCleanupUnrackedAdminBuildpacks(StagingStartRequestAdminBuildpack[] adminBuildpacks, string adminBuildpackDir)
```

Comment:
I wanted to mean untracked :)
But I'll rename it to TryCleanupUnusedAdminBuildpacks. "Try" meaning that it will skip the cleanup if a buildpack is in use.